tests/test_engine.py

```diff
@@ -76,6 +99,36 @@ def test_has_table(self):
         d = engine.has_table('mytable')
         assert self.successResultOf(d) is True
 
+    def test_pooled_connection_reuse(self):
+        table = Table('my_pool_table', sqlalchemy.MetaData(),
+                      Column('id', Integer, primary_key=True),
+                      Column('name', String, nullable=False))
+        data = [{'id': 0,
+                 'name': 'a'},
+                {'id': 1,
+                 'name': 'b'},
+                {'id': 2,
+                 'name': 'c'}]
+        # max_overflow kwarg broken, but defaults to 10
+        engine = create_engine()
+        engine.execute(CreateTable(table))
+        engine.execute(table.insert(), data)
+        pool_size = engine._engine.pool.size
+        if callable(pool_size):
+            pool_size = pool_size()
+        if isinstance(engine._engine.pool, sqlalchemy.pool.QueuePool):
+            pool_size += engine._engine.pool._max_overflow
+        arbitrary = Random()
+        for _ in range(pool_size + 1):
+            entry = arbitrary.choice(list(range(3)))
+            d = engine.execute(table.select().where(
+                table.c.id == data[entry]['id']))
+
+            result = self.successResultOf(d)
```

Comment:
Okay I think I'm starting to see the problem here.  You're using `FakeThreadedReactor`, which *doesn't actually use multiple threads*, to schedule work that is assumed to be *synchronous*, and then asserting that it completes immediately.  This is a more or less invalid use of `successResultOf`; there's real asynchrony going on when you're talking to a remote database.

I don't totally understand what the expected lifecycle of a connection is when you just keep calling `execute`; at what point is `recycle` supposed to get called?  It might be helpful for you to debug this with http://docs.sqlalchemy.org/en/rel_1_1/core/events.html#connection-pool-events in order to see if `checkin` is ever called.  I'm guessing that some of them remain checked out because of other failing tests that are still holding on to results that haven't been retrieved yet?

In any case, though, the way these tests are written is very confusing, since we're turning the async interface into a blocking interface again, but then testing it in a realistic async integration testing environment, without any real threads.  I don't know what the correct behavior should be, or what code paths this is even really supposed to exercise.